ARROW-10224: [Python] Add support for Python 3.9 except macOS wheel and Windows wheel

terencehonles opened the pull request:
Adds support and testing for Python 3.9. I am looking for review as this change may have touched too many things, but I'm also looking to get the CI to test all the different environments.

H/T: @kou, the documentation and #5685 for helping me get this off the ground.

terencehonles commented on dev/tasks/homebrew-formulae/apache-arrow.rb:
```diff
@@ -16,7 +16,7 @@ class ApacheArrow < Formula
   depends_on "openssl@1.1"
   depends_on "protobuf"
-  depends_on "python@3.8"
+  depends_on "python@3.9"
```
I'm assuming homebrew should be on the latest Python? It does look like it's available now https://formulae.brew.sh/formula/python@3.9

terencehonles commented on dev/tasks/python-wheels/win-build.bat:
```diff
@@ -110,6 +110,6 @@ set ARROW_TEST_DATA=%ARROW_SRC%\testing\data
 @rem run the python tests, but disable the cython because there is a linking
-@rem issue on python 3.8
+@rem issue on python 3.9
 set PYARROW_TEST_CYTHON=OFF
```
I don't know if this still applies, but I figured it probably would

terencehonles replied:
Sounds good to me, thanks for kicking things off. I had also create https://github.com/ursa-labs/ursabot/pull/205 to adjust buildbot since I realized 3.9 wasn't being built just by me creating this PR.

terencehonles commented on python/pyproject.toml:
```diff
@@ -19,7 +19,9 @@
 requires = [
     "cython >= 0.29",
     "numpy==1.14.5; python_version<'3.7'",
-    "numpy==1.16.0; python_version>='3.7'",
```
Files:

- python/pyproject.toml
- python/requirements-wheel-build.txt
- python/requirements-wheel-test.txt

Seemed to closely match, but `pyproject.toml` mentions a different version for Python 3.7, I pinned what was specified for Python 3.8, but it probably can accept `1.19.2` which I'm not sure if it's the first release to support Python 3.9, but it does build properly.

kou replied:
> but it probably can accept 1.19.2 which I'm not sure if it's the first release to support Python 3.9

How about confirming it with CI?

terencehonles replied:
Just to follow up, 1.19.2 did work for me, I'm just saying there might be a earlier release that works. It's probably simplest/best to use 1.19.2 since it's mentioned in their changelog:

> This release supports Python 3.6-3.8. Cython >= 0.29.21 needs to be used when building with Python 3.9 for testing purposes.
https://numpy.org/devdocs/release/1.19.2-notes.html

...looking at 1.19.1 it says the same thing so it might be safe to go backwards

terencehonles commented:
It looks like I "missed" the github workflows, but that might need to be done after the wheels are built and published as I haven't looked at what exactly are doing, and I kinda doubt they are building from source. I can open that up as a separate PR if that is helpful.

kou commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
It looks like cmake was already part of the quay.io/pypa/manylinux2014_x86_64:2020-10-06-b2ca7a1 and  `scripts/install_cmake.sh` needed to be updated.

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
Not sure if that will work for me, but might as well give it a shot 🤷

kou commented:
@github-actions crossbow submit -g conda -g wheel

kou commented:
(We need to wait a while until our comments are processed.)

terencehonles commented on dev/tasks/python-wheels/travis.osx.yml:
```diff
@@ -31,7 +31,7 @@ addons:
+      - python@3.9
       - thrift
       - wget
     update: false
```
I think it's possible Travis doesn't have Python 3.9 in it's homebrew cache, but I would expect a different error than it succeeds and then `/usr/local/opt/python@3.8/libexec/bin/pip` is just not found...

Nevermind! Apparently it's just a "warning" :unamused: :
> Warning: 'python@3.9' formula is unreadable: No available formula with the name "python@3.9" 

https://travis-ci.org/github/ursa-labs/crossbow/builds/733873746#L145

terencehonles commented:
I reverted some of the OS X changes. I'm not sure I actually needed to change Python 3.8 there since it looks like that travis file is templated out to all OS X wheels and it is still using Python 3.8 in all of them. That's either a bug or that there's something else that's using the `python_version` that's also specified in the travis jobs (more likely the latter).

pitrou commented:
Note that Numpy doesn't provide [pre-compiled wheels](https://pypi.org/project/numpy/#files) for 3.9 currently.

pitrou commented:
Also, conda-forge doesn't seem to be ready either.

terencehonles commented:
> Note that Numpy doesn't provide [pre-compiled wheels](https://pypi.org/project/numpy/#files) for 3.9 currently.
> 
> Also, conda-forge doesn't seem to be ready either.

Yeah, that's fine. I'm not really in a rush, I just wanted to help to get things in place that we could move forward as soon as all of the dependencies are met (and to identify where things are and what is still missing)

xhochy commented:
> Also, conda-forge doesn't seem to be ready either.

Should be ready now.

xhochy commented:
@github-actions crossbow submit -g conda

terencehonles commented:
> @github-actions crossbow submit -g conda

I _think_ the issue with the failed jobs is I didn't bump the numpy version in the `conda-recipies/.ci_support` ? I noticed that even Python 3.8 was set to numpy 1.16 so I might be wrong about that.

I updated the following https://github.com/apache/arrow/pull/8386/files#diff-0bd27b86c3660caa163f79bd5077954145a206949c81b9d8770a5d06a7c89c5bR39-R40 in each of the files

```yaml
numpy:
- '1.19.2'
```

and rebased the changes. It looks like a test is now failing and I'll look to see if I can quickly resolve that.

terencehonles commented:
I pinned semver to 2.10.2 since it looks like there is a breaking change for `archery/release.py` that was committed in the last few days. I can look to create a PR for that, but I figured it probably shouldn't be rolled into this PR.

terencehonles commented:
@xhochy do you mind if I get a resubmit to test the conda recipe changes?

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
I didn't realize the .ci_support files were vendored from https://github.com/conda-forge/arrow-cpp-feedstock. I updated the files to https://github.com/conda-forge/arrow-cpp-feedstock/commit/abbcc009d4171a8d6cb5dbfd5c49ddc38ab8e7f5

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
I'm definitely moving into very uncharted territory for me since I'm not familiar with conda, but I believe it looks like numpy for Py39 was pinned to 1.16 here [dev/tasks/conda-recipes/arrow-cpp/meta.yaml](https://github.com/apache/arrow/pull/8386/files#diff-e1dc9dc44f82817c013424829588fc3fe48e63c78b42ce47e012016a243de09bL182) and I am expecting that removing that should fix Py39 since looking at https://dev.azure.com/ursa-labs/crossbow/_build/results?buildId=19036&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181&l=711 it looks like Py39 is still trying to use numpy 1.16 instead of 1.19 as specified in the .ci_support files for Py3.9.

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

xhochy commented:
Conda builds are fine, we sadly seem to be limited on drone.io and thus we should ignore the `-aarch64` builds for now.

terencehonles commented:
Py3.7 looks like it's something wrong when testing the wheel (I'll see if I can reproduce locally). I and I'm not sure what do do about llvm 10 [here](https://github.com/apache/arrow/pull/8386/checks?check_run_id=1282217190)

terencehonles commented:
> Py3.7 looks like it's something wrong when testing the wheel (I'll see if I can reproduce locally). I and I'm not sure what do do about llvm 10/11 [here](https://github.com/apache/arrow/pull/8386/checks?check_run_id=1282217190)

I'll check if it's just a pinning issue since it looks related to https://github.com/apache/arrow/pull/8492

terencehonles commented:
> > Py3.7 looks like it's something wrong when testing the wheel (I'll see if I can reproduce locally). I and I'm not sure what do do about llvm 10/11 [here](https://github.com/apache/arrow/pull/8386/checks?check_run_id=1282217190)
> 
> I'll check if it's just a pinning issue since it looks related to #8492

It looks like removing the unpinned llvm and jusing using the pinned llvm@8 did the trick. In the long term it makes sense to try to support LLVM 11 since it doesn't look like homebrew is looking to support llvm@10 https://github.com/Homebrew/homebrew-core/pull/63177

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
I believe I identified the test on Python3.7 that is failing, and going to see if that fixes things

@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
> I believe I identified the test on Python3.7 that is failing (same as #3403), and going to see if that fixes things

Looks like there was an existing typo and I broke a test when I changed (numpy is 1.16.0 not 0.16.0):

```diff
diff --git a/python/pyarrow/tests/test_pandas.py b/python/pyarrow/tests/test_pandas.py
index f46e75fc3..44b62a95f 100644
--- a/python/pyarrow/tests/test_pandas.py
+++ b/python/pyarrow/tests/test_pandas.py
@@ -2787,7 +2789,7 @@ def _check_serialize_components_roundtrip(pd_obj):
         tm.assert_series_equal(pd_obj, deserialized)
 
 
-@pytest.mark.skipif(LooseVersion(np.__version__) >= '0.16',
+@pytest.mark.skipif('0.16.0' <= LooseVersion(np.__version__) < '0.16.1',
                     reason='Until numpy/numpy#12745 is resolved')
 def test_serialize_deserialize_pandas():
     # ARROW-1784, serialize and deserialize DataFrame by decomposing
```

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
I can't tell what's actually the issue with the OS X Python 3.9 builds. I see https://travis-ci.com/github/ursa-labs/crossbow/builds/191336854 and my guess is that CMake [has not been updated in either homebrew or on Travis](https://travis-ci.com/github/ursa-labs/crossbow/builds/191336854#L119) to support [Python 3.9](https://travis-ci.com/github/ursa-labs/crossbow/builds/191336854#L467-L475)?

terencehonles commented:
@kou not sure if you have any thoughts on the following:

- For the windows wheel it looks like the issue is the VS runtime is pinned [here](https://github.com/apache/arrow/blob/fa456de6d03c4fa5bac104b5b31f72d7be92f450/dev/tasks/python-wheels/win-build.bat#L28). I'm not sure if I can just remove that or what I should adjust it to.
- For Mac it looks like it's [something to do with CMake](https://travis-ci.com/github/ursa-labs/crossbow/builds/191503619#L467-L475) and I'm not sure where to go from here
- C++ / AMD64 Windows MinGW 64 C++ (pull_request) looks like it's starting to fail now, and it's because of a timeout error. The timeout is set to 5 minutes, but I'm wondering if it's OK for me to bump it up higher.

kou commented:
>     * For the windows wheel it looks like the issue is the VS runtime is pinned [here](https://github.com/apache/arrow/blob/fa456de6d03c4fa5bac104b5b31f72d7be92f450/dev/tasks/python-wheels/win-build.bat#L28). I'm not sure if I can just remove that or what I should adjust it to.

Could you try removing `"vs2015_runtime<14.16"`?

@kszucs Do you remember why you added it in #7129 ? 

>     * For Mac it looks like it's [something to do with CMake](https://travis-ci.com/github/ursa-labs/crossbow/builds/191503619#L467-L475) and I'm not sure where to go from here

Could you try the command lines that are executed in CMake on the environment?

  * https://gitlab.kitware.com/cmake/cmake/-/blob/v3.17.0/Modules/FindPython/Support.cmake#L2333
  * https://gitlab.kitware.com/cmake/cmake/-/blob/v3.17.0/Modules/FindPython/Support.cmake#L2357

>     * C++ / AMD64 Windows MinGW 64 C++ (pull_request) looks like it's starting to fail now, and it's because of a timeout error. The timeout is set to 5 minutes, but I'm wondering if it's OK for me to bump it up higher.

Generally, a timeout error in C++ test means the test process is stuck. It can't be fixed by increasing timeout.

terencehonles commented:
@github-actions crossbow submit wheel-win-*

terencehonles commented:
@github-actions crossbow submit wheel-osx-*

terencehonles commented:
@github-actions crossbow submit wheel-osx-*

terencehonles commented:
> > ```
> > * For the windows wheel it looks like the issue is the VS runtime is pinned [here](https://github.com/apache/arrow/blob/fa456de6d03c4fa5bac104b5b31f72d7be92f450/dev/tasks/python-wheels/win-build.bat#L28). I'm not sure if I can just remove that or what I should adjust it to.
> > ```
> 
> Could you try removing `"vs2015_runtime<14.16"`?

It looks like all of the windows wheels are failing on

```
--   Including C:/Miniconda3-x64/envs/wheel-build/Library/lib/cmake/boost_regex-1.74.0/libboost_regex-variant-shared.cmake
--   [ ] boost_regex.lib
CMake Error at C:/Miniconda3-x64/envs/wheel-build/Library/lib/cmake/Boost-1.74.0/BoostConfig.cmake:141 (find_package):
  Found package configuration file:
    C:/Miniconda3-x64/envs/wheel-build/Library/lib/cmake/boost_regex-1.74.0/boost_regex-config.cmake
  but it set boost_regex_FOUND to FALSE so package "boost_regex" is
  considered to be NOT FOUND.  Reason given by package:
  No suitable build variant has been found.
  The following variants have been tried and rejected:
  * boost_regex.lib (shared, Boost_USE_STATIC_LIBS=ON)
Call Stack (most recent call first):
```

wesm commented:
We need to add 1.74.0 to https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake#L770

terencehonles commented:
> We need to add 1.74.0 to https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake#L770

Thanks! I was literally just looking at that file. I'll adjust. What about python/CMakeLists.txt?

terencehonles commented:
@github-actions crossbow submit wheel-win-*

terencehonles commented:
> Could you try the command lines that are executed in CMake on the environment?
> 
> * https://gitlab.kitware.com/cmake/cmake/-/blob/v3.17.0/Modules/FindPython/Support.cmake#L2333
> * https://gitlab.kitware.com/cmake/cmake/-/blob/v3.17.0/Modules/FindPython/Support.cmake#L2357

It looks like the issue is likely this https://github.com/numpy/numpy/issues/15947 when using Python directly as suggested:

```
Building wheels for collected packages: numpy
  Building wheel for numpy (PEP 517): started
  Building wheel for numpy (PEP 517): still running...
  Building wheel for numpy (PEP 517): still running...
  Building wheel for numpy (PEP 517): finished with status 'done'
  Created wheel for numpy: filename=numpy-1.19.2-cp39-cp39-macosx_10_13_x86_64.whl size=4626041 sha256=be42ba9492a0f018618368d2193572162468dd640733db26582bae57715247d5
  Stored in directory: /Users/travis/Library/Caches/pip/wheels/a3/17/dd/f2dba23a35bb6008732772ccfb13d3d0e537fbc6919ce6862b
Successfully built numpy
Installing collected packages: cython, numpy
Successfully installed cython-0.29.21 numpy-1.19.2
++python -c 'import numpy; print(numpy.get_include())'
init_dgelsd failed init
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/travis/build/ursa-labs/crossbow/venv/lib/python3.9/site-packages/numpy/__init__.py", line 286, in <module>
    raise RuntimeError(msg)
RuntimeError: Polyfit sanity test emitted a warning, most likely due to using a buggy Accelerate backend. If you compiled yourself, see site.cfg.example for information. Otherwise report this to the vendor that provided NumPy.
RankWarning: Polyfit may be poorly conditioned
```

terencehonles commented:
@github-actions crossbow submit wheel-osx-*

terencehonles commented:
@github-actions crossbow submit wheel-osx-mavericks-cp39

terencehonles commented:
@github-actions crossbow submit wheel-osx-mavericks-cp39

terencehonles commented:
@github-actions crossbow submit wheel-osx-mavericks-cp39

terencehonles commented:
@github-actions crossbow submit wheel-osx-mavericks-cp39

terencehonles commented:
@github-actions crossbow submit wheel-osx-mavericks-cp39

terencehonles commented:
@github-actions crossbow submit wheel-osx-mavericks-cp39

terencehonles commented:
@github-actions crossbow submit wheel-osx-mavericks-cp39

terencehonles commented:
@github-actions crossbow submit wheel-osx-mavericks-cp39

terencehonles commented:
> Revision: [26f3c31](https://github.com/apache/arrow/commit/26f3c310230c6ef04471fe83f87f0b806aa5d551)
> 
> Submitted crossbow builds: [ursa-labs/crossbow @ actions-675](https://github.com/ursa-labs/crossbow/branches/all?query=actions-675)
> 
> Task	Status
> wheel-osx-mavericks-cp39	[![TravisCI](https://camo.githubusercontent.com/30faf412fbc86c862154793b5f3fab9242c664a0/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637352d7472617669732d776865656c2d6f73782d6d6176657269636b732d637033392e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)

There we go!

terencehonles commented:
@github-actions crossbow submit wheel-osx-*

terencehonles commented:
> Revision: [d26b20c](https://github.com/apache/arrow/commit/d26b20cb64debc6f022fdce1cde3ea3f812981ad)
> 
> Submitted crossbow builds: [ursa-labs/crossbow @ actions-676](https://github.com/ursa-labs/crossbow/branches/all?query=actions-676)
> 
> Task	Status
> wheel-osx-high-sierra-cp35m	[![TravisCI](https://camo.githubusercontent.com/3726b981c75aa919a2a31fd7060c5e6dbf80000a/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d686967682d7369657272612d637033356d2e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-high-sierra-cp36m	[![TravisCI](https://camo.githubusercontent.com/5c2adaeb2826965478066f4359d34ad392d589eb/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d686967682d7369657272612d637033366d2e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-high-sierra-cp37m	[![TravisCI](https://camo.githubusercontent.com/9c99675bbb942dff44c23fefaca7b55e06cb165c/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d686967682d7369657272612d637033376d2e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-high-sierra-cp38	[![TravisCI](https://camo.githubusercontent.com/8aa78922486d05cb39cd23728422eabd8d4b0975/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d686967682d7369657272612d637033382e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-high-sierra-cp39	[![TravisCI](https://camo.githubusercontent.com/fac0c414284bf2dc05de67078d2fc8534b78b68b/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d686967682d7369657272612d637033392e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-mavericks-cp35m	[![TravisCI](https://camo.githubusercontent.com/b73fe248b67efe4bdeb28af32ca67b5d99ef645a/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d6d6176657269636b732d637033356d2e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-mavericks-cp36m	[![TravisCI](https://camo.githubusercontent.com/83a61b7877d12c6b3c160c7bbf50c50986594fcc/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d6d6176657269636b732d637033366d2e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-mavericks-cp37m	[![TravisCI](https://camo.githubusercontent.com/b24716b5df6fbe2dde764790841d3f11fd7152c8/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d6d6176657269636b732d637033376d2e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-mavericks-cp38	[![TravisCI](https://camo.githubusercontent.com/4aa0c1f5735c8dedd27ad59d48652635403d11f9/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d6d6176657269636b732d637033382e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)
> wheel-osx-mavericks-cp39	[![TravisCI](https://camo.githubusercontent.com/d14c6f5667aab0749a24d701c72a6ce0661fbdbe/68747470733a2f2f696d672e736869656c64732e696f2f7472617669732f757273612d6c6162732f63726f7373626f772f616374696f6e732d3637362d7472617669732d776865656c2d6f73782d6d6176657269636b732d637033392e737667)](https://travis-ci.org/ursa-labs/crossbow/branches)

Looks like more patience is needed. Since numpy and pandas will be built from source they exceed the time limit on Travis :confused: , I'll follow up again once at least one of them has Python 3.9 wheels.

sir-sigurd commented:
> Looks like more patience is needed. Since numpy and pandas will be built from source they exceed the time limit on Travis confused , I'll try to follow up again once at least one of them has Python 3.9 wheels.

numpy uploaded Python 3.9 wheels yesterday: https://pypi.org/project/numpy/1.19.3/#files

manselmi commented:
> > Looks like more patience is needed. Since numpy and pandas will be built from source they exceed the time limit on Travis confused , I'll try to follow up again once at least one of them has Python 3.9 wheels.
> 
> numpy uploaded Python 3.9 wheels yesterday: https://pypi.org/project/numpy/1.19.3/#files

pandas also has published Python 3.9 wheels to PyPI: https://pypi.org/project/pandas/#files

terencehonles commented:
Right on! I'll look to see if I can pick this back up :tada:

terencehonles commented:
Not sure what's up with appveyor or the failed tests, but I'm kicking them off again just in case they are flakey.

terencehonles commented:
@github-actions crossbow submit wheel-win-*

terencehonles commented:
> Revision: [3d8adc2](https://github.com/apache/arrow/commit/3d8adc29623abae7e2cb29e5b2a77cc3161c12e8)
> 
> Submitted crossbow builds: [ursa-labs/crossbow @ actions-687](https://github.com/ursa-labs/crossbow/branches/all?query=actions-687)
> 
> Task	Status
> wheel-win-cp35m	[![Appveyor](https://camo.githubusercontent.com/7748e1e2bc60fc20518cd1ba820a8cdaab17f6d0/68747470733a2f2f696d672e736869656c64732e696f2f6170707665796f722f63692f757273612d6c6162732f63726f7373626f772f616374696f6e732d3638372d6170707665796f722d776865656c2d77696e2d637033356d2e737667)](https://ci.appveyor.com/project/ursa-labs/crossbow/history)
> wheel-win-cp36m	[![Appveyor](https://camo.githubusercontent.com/18d8a299a85d1d6bc730daced015dab948bb0804/68747470733a2f2f696d672e736869656c64732e696f2f6170707665796f722f63692f757273612d6c6162732f63726f7373626f772f616374696f6e732d3638372d6170707665796f722d776865656c2d77696e2d637033366d2e737667)](https://ci.appveyor.com/project/ursa-labs/crossbow/history)
> wheel-win-cp37m	[![Appveyor](https://camo.githubusercontent.com/db9603a5cfe4053b72314b58acb838af16e6ba11/68747470733a2f2f696d672e736869656c64732e696f2f6170707665796f722f63692f757273612d6c6162732f63726f7373626f772f616374696f6e732d3638372d6170707665796f722d776865656c2d77696e2d637033376d2e737667)](https://ci.appveyor.com/project/ursa-labs/crossbow/history)
> wheel-win-cp38	[![Appveyor](https://camo.githubusercontent.com/244bedbba63aa8f25c0547b4b3ab15bc7773fc3e/68747470733a2f2f696d672e736869656c64732e696f2f6170707665796f722f63692f757273612d6c6162732f63726f7373626f772f616374696f6e732d3638372d6170707665796f722d776865656c2d77696e2d637033382e737667)](https://ci.appveyor.com/project/ursa-labs/crossbow/history)
> wheel-win-cp39	[![Appveyor](https://camo.githubusercontent.com/7969077db0bd89e4902013665d30c58b332fcbb7/68747470733a2f2f696d672e736869656c64732e696f2f6170707665796f722f63692f757273612d6c6162732f63726f7373626f772f616374696f6e732d3638372d6170707665796f722d776865656c2d77696e2d637033392e737667)](https://ci.appveyor.com/project/ursa-labs/crossbow/history)

@kou any chance you know what happened to appveyor?

kou commented:
It seems that https://ci.appveyor.com/project/ursa-labs/crossbow doesn't exist.

@kszucs Could you enable AppVeyor for https://github.com/ursa-labs/crossbow ?

kszucs commented:
@kou crossbow has been deleted from appveyor and I'm banned to do anything. Support has not responded yet. You may try to setup the appveyor integration for that repository, perhaps it would work for your user. Otherwise we can switch to another crossbow repository or port the windows wheel builds to another CI.

kou commented:
I see. It seems that our Crossbow + AppVeyor usage isn't suitable for AppVeyor. I will port the Crossbow tasks to GitHub Actions from AppVeyor.

kszucs commented:
That would be great, thanks Kou!

kou commented:
Sorry... We can't use GitHub Actions for building wheels for Python on Windows.
It requires Visual Studio 2015 but GitHub Actions provides only 2017 and 2019.

pitrou commented:
Hmm... why is VS2015 required?

pitrou commented:
I'll note that [cibuildwheel](https://github.com/joerick/cibuildwheel) supports building Windows wheels on GHA.  The [setup](https://cibuildwheel.readthedocs.io/en/stable/setup/#github-actions) simply uses the `windows-latest` image.

pitrou commented:
Also, in case you are worrying about ABI issues, they are gone with the [UCRT](https://docs.microsoft.com/en-us/cpp/porting/upgrade-your-code-to-the-universal-crt?view=msvc-160).

> The UCRT supports a stable ABI based on C calling conventions, and it conforms closely to the ISO C99 standard, with only a few exceptions. **It is no longer tied to a specific version of the compiler.** [emphasis mine]

kou commented:
Oh, I didn't know about it.
I thought we need to use the same Visual Studio that is used for building Python.
I'll try using Visual Studio 2017.

kou commented:
@terencehonles Could you rebase on master to use #8620?

terencehonles commented:
done

kou commented:
@github-actions crossbow submit wheel-win-*

Thanks!

kou commented:
@github-actions crossbow submit wheel-win-*

kou commented:
@github-actions crossbow submit wheel-win-*

kou commented:
@github-actions crossbow submit wheel-win-*

kou commented:
@github-actions crossbow submit wheel-win-cp38

pitrou commented:
Looks like there's no Github Actions config here: https://github.com/ursa-labs/crossbow/tree/actions-698

kou commented:
It's not a problem that `actions-XXX` branch doesn't have GitHub Actions config.
It's a problem that `actions-XXX-github-YYY` branch isn't created.

It seems that we need to add the `workflow` scope to GitHub Personal Access Token for Crossbow.
@kszucs Could you add the `workflow` scope to the Personal Access Token registered by https://issues.apache.org/jira/browse/INFRA-19954 ?

kszucs commented:
> It's not a problem that `actions-XXX` branch doesn't have GitHub Actions config.
> It's a problem that `actions-XXX-github-YYY` branch isn't created.
Correct.
> 
> It seems that we need to add the `workflow` scope to GitHub Personal Access Token for Crossbow.
> @kszucs Could you add the `workflow` scope to the Personal Access Token registered by https://issues.apache.org/jira/browse/INFRA-19954 ?
Is this newly introduced by github? I'm going to take a look why the task's branch is not pushed.

kszucs commented:
@github-actions crossbow submit wheel-win-cp38

kszucs commented:
@github-actions crossbow submit wheel-win-cp38

kszucs commented:
@github-actions crossbow submit wheel-win-cp38

kszucs commented:
@kou enabled the workflow scope and it works now.

kou commented:
Thanks!

> Is this newly introduced by github?

It seems that it's introduced at 2020-06-03: https://github.blog/changelog/2020-06-03-personal-access-tokens-workflow-permission/

GitHub may disable the following care recently:

> To prevent breaking existing workflows we are enabling this scope on all existing PATs.

kou commented:
@github-actions crossbow submit wheel-win-*

kou commented:
@github-actions crossbow submit wheel-win-cp39

kou commented:
@github-actions crossbow submit wheel-win-cp39

kou commented:
@github-actions crossbow submit wheel-win-cp39

kou commented:
@github-actions crossbow submit wheel-win-cp39

kou commented:
@github-actions crossbow submit wheel-win-cp39

kou commented:
@github-actions crossbow submit wheel-win-cp39

kou commented:
@github-actions crossbow submit wheel-win-cp39

kou commented:
@github-actions crossbow submit wheel-win-cp39

kou commented:
The current status:

  * We can't use old packages that depend on `vs2015_runtime<14.16` on conda-forge for Python 3.9 on Windows.
    Because Python 3.9 for Windows on conda-forge depend on `vs2015_runtime 14.16.27012`: https://anaconda.org/conda-forge/python/files/modal/info/5f8784bea58f0d0a5480e400 

  * We can't use recent packages that depend on `vs2015_runtime>=14.16` on conda-forge to build wheels for Windows.
    Because recent packages on conda-forge don't provide static libraries: https://github.com/conda-forge/cfep/pull/34

Ideas for the next step:

  1. We use our bundle mechanism to build all dependencies instead of depending on libraries provided by conda-forge.

  2. We use other packaging system like [vcpkg](https://github.com/microsoft/vcpkg) to collect dependencies.

I tried 1 and found that our gRPC build configuration has some problems with Visual C++. For example:

  * RE2 dependency is missing. (We need RE2 for not only Gandiva but also Flight.)
  * CMake options to find bundled zlib is wrong. (We should use `-DZLIB_INCLUDE_DIR` and `-DZLIB_LIBRARY` instead of `-DCMAKE_PREFIX_PATH`.)
  * It seems that dynamic library and static library are mixed unexpectedly: https://github.com/ursa-labs/crossbow/runs/1394393091?check_suite_focus=true
  * ...

I didn't try 2 yet.

I don't have enough time to fix this problem for now. I may work on this again next month but others can work on this.

stonebig commented:
hi, is there a dev Python-3.9 wheel for Windows "pip" already produced by the build system ?

kou commented:
Not yet.

ns-admetrics commented:
Arch Linux switched to Python 3.9.0 today (which means that Python 3.8 was removed from that distribution's repositories).

I was able to build a wheel for both `origin/HEAD` and `tags/apache-arrow-2.0.0` by following https://github.com/apache/arrow/blob/master/docs/source/developers/python.rst to the letter, which leads me to believe there is not much if any actual incompatibility nor issue with Python 3.9.0.

Would it perhaps be possible to build and publish Linux wheels to PyPI? Remarks above imply there are some bigger issues on the Windows platform.

terencehonles commented:
I can rebase if needed (I had done that first), but figured it might be easier if others are committing to just merge. It's going to be squashed anyways so I doubt that matters too much.

terencehonles commented:
@xhochy not sure if your debugging in #8778 changes anything as far as the things attempted so far in @kou 's message https://github.com/apache/arrow/pull/8386#issuecomment-728321361

Unfortunately I don't have access to a Windows machine right now and don't really have the time to set up a VM or borrow one in order to test things out locally. I'm not sure if releasing wheels - windows is an option as @ns-admetrics suggests.

xhochy commented:
I got a bit further in #8778 but we aren't fully there yet. I would rather merge the changes here bit-by-bit as the PR got already quite a bit stale.

@terencehonles would issue the subparts as a PR or should I take care of that?

terencehonles commented:
> I got a bit further in #8778 but we aren't fully there yet. I would rather merge the changes here bit-by-bit as the PR got already quite a bit stale.
> 
> @terencehonles would issue the subparts as a PR or should I take care of that?

I can probably handle that. Can I just pull out the windows wheels or how do you think it should be broken up?

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
@github-actions crossbow submit -g conda -g wheel

terencehonles commented:
I _believe_ https://github.com/apache/arrow/pull/8386/commits/ec5e6797e48fb0f4679c12496444ceead0861524 should be sufficient and I shouldn't have to remove `dev/tasks/conda-recipes/.ci_support/win*`. It looks like crossbow agrees, but not sure if there is any other builders that might be looking elsewhere.

terencehonles commented:
@github-actions crossbow submit -g wheel

terencehonles commented:
I'll try to build locally, there seems to be a SSL/TLS error when trying to use `yum`

terencehonles commented:
@github-actions crossbow submit wheel-manylinux*

terencehonles commented:
@xhochy is windows failing on `master`? I see your PR #8881 and not sure if I should leave windows as is. I also notice that the osx wheels aren't building either and not sure if that needs to be moved to github from travis too (as windows was).

terencehonles commented:
@github-actions crossbow submit wheel-manylinux1-cp36m wheel-manylinux1-cp38

terencehonles commented:
That's odd that yum worked on one and not the other. It looks like it might be a tempermental issue and the fix in #8881 may need to be applied to manylinux1? It looks like that's on another tree and _might_ be related.

https://github.com/apache/arrow/pull/8386#issuecomment-742112004
![Screenshot from 2020-12-10 12-33-57](https://user-images.githubusercontent.com/438813/101826636-1f7a5f00-3ae4-11eb-98a6-8790e291c6c1.png)

https://github.com/apache/arrow/pull/8386#issuecomment-742219891
![Screenshot from 2020-12-10 12-34-16](https://user-images.githubusercontent.com/438813/101826646-243f1300-3ae4-11eb-877f-6c37d7be2525.png)

https://github.com/apache/arrow/pull/8386#issuecomment-742740508
![Screenshot from 2020-12-10 12-34-42](https://user-images.githubusercontent.com/438813/101826657-286b3080-3ae4-11eb-91cc-b7ae65e2e33d.png)

terencehonles commented:
@github-actions crossbow submit wheel-manylinux1-*

terencehonles commented:
removing the httpS from the change, and that looks like it fixes the problem. I'm going to push this change upstream, but not sure if this PR should be blocked on that.

terencehonles commented:
@github-actions crossbow submit wheel-manylinux1-*

terencehonles commented:
Looks like that did the trick, I can create a comment linking to https://github.com/pypa/manylinux/pull/872 or we can wait till that is merged in and built

terencehonles commented:
It looks like the windows wheels are broken on `master` https://github.com/apache/arrow/pull/8878#issuecomment-742417983 and not sure if the OSX packages should be updated to use Github Actions rather than travis which seems to not be running.

Other than waiting on https://github.com/pypa/manylinux/pull/872 should I do anything futher @xhochy / @kou ?

kou commented on cpp/Brewfile:
```diff
@@ -25,7 +25,6 @@ brew "flatbuffers"
 brew "git"
 brew "glog"
 brew "grpc"
-brew "llvm"
```
Can we revert this change?

terencehonles replied:
Do we actually want brew `llvm` and `llvm@8`? Reverting anyways.

kou replied:
`llvm@8` is just for linting. We use `clang-format` and so on.
`llvm` is for Gandiva.

kou commented on dev/release/verify-release-candidate.sh:
```diff
@@ -656,7 +656,7 @@ test_linux_wheels() {
 
 test_macos_wheels() {
-  local py_arches="3.5m 3.6m 3.7m 3.8"
+  local py_arches="3.5m 3.6m 3.7m 3.8 3.9"
```
Could you revert this and other macOS wheel support related changes?
Let's work on macOS wheel support in separated pull request.

kou commented:
@github-actions crossbow submit wheel-manylinux* conda-*

kou commented on dev/tasks/python-wheels/github.win.yml:
```diff
@@ -53,9 +53,9 @@ jobs:
+        shell: cmd
         run: |
-          arrow/dev/tasks/python-wheels/win-build.bat
+          call arrow/dev/tasks/python-wheels/win-build.bat
```
Sorry, could you revert this change by me? https://github.com/apache/arrow/pull/8386/commits/5e03e010bc65308452c95d51ed21f0baebebf059
This was meaningless.

terencehonles replied:
Ok, I left it in because I wasn't sure if it actually allowed the error code to be used, I'll revert since it sounds like it wasn't a problem

kou reviewed (commented):
Let's work on macOS wheel and Windows wheel support as another tasks.
Could you open issues for them?

Could you also revert `testing/` changes?

kou commented:
@xhochy Can we accept `conda-linux-gcc-py*-aarch64` failures in this pull request?

xhochy commented:
> @xhochy Can we accept `conda-linux-gcc-py*-aarch64` failures in this pull request?

Yes. CI is broken but I know that upstream they work. So we're fine here.

terencehonles commented:
Testing to see if the failed checks could be bumped to passing.

terencehonles commented:
Looks like it's probably not going to pass, I'll see if I can look into it later. I got some feedback on https://github.com/pypa/manylinux/pull/872 and I can either update my changes here or maybe that will get merged in soon.

kou commented:
Do you mean the failures of MinGW related jobs?
They aren't related to this pull request. We can ignore them in this pull request.

It seems that pypa/manylinux#872 has been merged. Do you want to update this pull request before we merge this?

terencehonles commented:
> Do you mean the failures of MinGW related jobs?
> They aren't related to this pull request. We can ignore them in this pull request.
> 
> It seems that [pypa/manylinux#872](https://github.com/pypa/manylinux/pull/872) has been merged. Do you want to update this pull request before we merge this?

yup, doing that now

terencehonles commented:
@github-actions crossbow submit wheel-manylinux* conda-*

kou reviewed (approved):
+1

Thanks! I'll merge this.

Could you open  JIRA issues for macOS wheel support and Windows wheel support?

idanw206 commented:
@kou When will the binaries be available in pypi? Thanks!

kou commented:
We'll release them in 2021-01.

n8henrie commented:
@kou You're likely well aware, but if looking for e.g. `manylinux2014` for 3.9:

```console
$ curl -s 'https://api.github.com/repos/ursa-labs/crossbow/releases' | jq -r '.[]["assets"] | .[]["browser_download_url"]' | grep manylinux2014-cp39
https://github.com/ursa-labs/crossbow/releases/download/nightly-2020-12-18-0-azure-wheel-manylinux2014-cp39/pyarrow-2.1.0.dev412-cp39-cp39-manylinux2014_x86_64.whl
https://github.com/ursa-labs/crossbow/releases/download/nightly-2020-12-18-0-azure-wheel-manylinux2014-cp39/pyarrow-2.1.0.dev412.tar.gz
```

Looks like the api [has a limit of 100 results per call](https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#list-releases), which doesn't go back far enough. Instead of dealing with pagination, it seems like `gh` may have a higher limit:

```console
$ gh release list --repo=ursa-labs/crossbow --limit=1000 | grep manylinux2014-c
p39
nightly-2020-12-18-0-azure-wheel-manylinux2014-cp39             nightly-2020-12-18-0-azure-wheel-manylinux2014-cp39     2020-12-18T07:26:20Z
nightly-2020-12-17-0-azure-wheel-manylinux2014-cp39             nightly-2020-12-17-0-azure-wheel-manylinux2014-cp39     2020-12-17T07:33:01Z
actions-807-azure-wheel-manylinux2014-cp39              actions-807-azure-wheel-manylinux2014-cp39      2020-12-17T01:23:13Z
actions-804-azure-wheel-manylinux2014-cp39              actions-804-azure-wheel-manylinux2014-cp39      2020-12-16T23:00:37Z
actions-799-azure-wheel-manylinux2014-cp39              actions-799-azure-wheel-manylinux2014-cp39      2020-12-16T08:54:37Z
nightly-2020-12-16-0-azure-wheel-manylinux2014-cp39             nightly-2020-12-16-0-azure-wheel-manylinux2014-cp39     2020-12-16T06:58:23Z
actions-794-azure-wheel-manylinux2014-cp39              actions-794-azure-wheel-manylinux2014-cp39      2020-12-15T21:35:42Z
nightly-2020-12-15-0-azure-wheel-manylinux2014-cp39             nightly-2020-12-15-0-azure-wheel-manylinux2014-cp39     2020-12-15T06:22:33Z
nightly-2020-12-14-0-azure-wheel-manylinux2014-cp39             nightly-2020-12-14-0-azure-wheel-manylinux2014-cp39     2020-12-14T07:19:20Z
actions-783-azure-wheel-manylinux2014-cp39              actions-783-azure-wheel-manylinux2014-cp39      2020-12-12T22:57:41Z
actions-772-azure-wheel-manylinux2014-cp39              actions-772-azure-wheel-manylinux2014-cp39      2020-12-11T07:19:59Z
actions-767-azure-wheel-manylinux2014-cp39              actions-767-azure-wheel-manylinux2014-cp39      2020-12-10T20:02:41Z
actions-759-azure-wheel-manylinux2014-cp39              actions-759-azure-wheel-manylinux2014-cp39      2020-12-10T04:27:52Z
actions-758-azure-wheel-manylinux2014-cp39              actions-758-azure-wheel-manylinux2014-cp39      2020-12-09T23:30:31Z
```

From the name of the release you can construct the URL:

```plaintext
https://github.com/ursa-labs/crossbow/releases/actions-783-azure-wheel-manylinux2014-cp39
```

And from there you can find:

```plaintext
https://github.com/ursa-labs/crossbow/releases/download/actions-783-azure-wheel-manylinux2014-cp39/pyarrow-2.1.0.dev363-cp39-cp39-manylinux2014_x86_64.whl
```

From here you can install directly with pip:

```console
$ python -m pip install https://github.com/ursa-labs/crossbow/releases/download/actions-783-azure-wheel-manylinux2014-cp39/pyarrow-2.1.0.dev363-cp39-cp39-manylinux2014_x86_64.whl
```

terencehonles commented:
Please be aware that's like running a nightly release, but if you are fine with that you can install a crossbow wheel. However, you probably want to just follow the instructions here https://arrow.apache.org/docs/python/install.html#installing-nightly-packages 

TL;DR:

> ```bash
> pip install --extra-index-url https://pypi.fury.io/arrow-nightlies/ \
>     --pre pyarrow
> ```

n8henrie commented:
Thanks for the suggestion!

Unfortunately my distro has already moved to 3.9 (including all my cuda and DL libraries etc) so I need a 39 compatible wheel or I can't use it. Nightly it is for now! This way I can make sure the release I'm installing is at least passing CI.

liyucheng09 commented:
Using the command mentioned by @terencehonles doesn't work for me.
```
(.venv) liyucheng@liyuchengdeMacBook-Air comments_generation % pip install --extra-index-url https://pypi.fury.io/arrow-nightlies/ --no-cache-dir --pre pyarrow
Looking in indexes: https://mirrors.aliyun.com/pypi/simple/, https://pypi.fury.io/arrow-nightlies/
Collecting pyarrow
  Downloading https://mirrors.aliyun.com/pypi/packages/fd/b7/78115614c4b227796cc87fff907930f6ae6dd999c5000d3d6ae5c2e54582/pyarrow-2.0.0.tar.gz (58.9 MB)
     |████▍                           | 8.2 MB 788 kB/s eta 0:01:05^C
```
This line still looking for pyarrow from the default indexes.

kou commented:
Adding `--verbose` to `pip` command line may help you.

liyucheng09 commented:
@kou Thanks for your help. I think it may because the link `https://pypi.fury.io/arrow-nightlies/` still lacks wheels for python3.9 on macos.

kou commented:
We're still working on it at #8915.

terencehonles commented:
> @kou Thanks for your help. I think it may because the link `https://pypi.fury.io/arrow-nightlies/` still lacks wheels for python3.9 on macos.

@liyucheng09 yes that's correct, if you're impatient and OK with using a crossbow build as @n8henrie was attempting you can use https://github.com/ursa-labs/crossbow/branches/all?query=actions-686 which built both mavericks and high sierra as you can see in https://github.com/apache/arrow/pull/8386#issuecomment-719687857

liyucheng09 commented:
@terencehonles Thanks a lot. I installed pyarrorw sucessfully.

I was just curious that when I type `pip install pyarrow` the error log shows that it was stopped during install numpy, which is already installed in my environment (maybe the version is too advanced?).

I thought that figure it out may help me to understand deeper about python package installing process.

terencehonles commented:
> @terencehonles Thanks a lot. I installed pyarrorw sucessfully.
> 
> I was just curious that when I type `pip install pyarrow` the error log shows that it was stopped during install numpy, which is already installed in my environment (maybe the version is too advanced?).
> 
> I thought that figure it out may help me to understand deeper about python package installing process.

I'm assuming you mean trying to install the most recently released pyarrow before you installed via crossbow? That's because pyarrow has a source distribution and you're trying to build from source. [Pip will not have installed numpy at install time](https://pip.pypa.io/en/stable/reference/pip/#pep-517-and-518-support) and is working to make isolated installs the default, if you have numpy already installed the version you have _may_ be an issue, but I think it's more likely it's too old not too new. You can try using the flag `--no-use-pep517` when running `pip install pyarrow` to see if your system has all the dependencies needed to build pyarrow (likely not, and that's why the CI is building it). Optionally [you can make pip more verbose](https://pip.pypa.io/en/stable/reference/pip/#cmdoption-v) with `-v`.

pyarrow is probably not the easiest to learn about Python's build process on, but you should read [their documentation](https://arrow.apache.org/docs/developers/python.html#python-development) if you would like to actually be able to build it from source.
